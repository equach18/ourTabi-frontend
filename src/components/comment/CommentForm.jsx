// CommentForm.jsx
import { useState } from "react";

function CommentForm({ addComment }) {
  const [text, setText] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();
    if (!text.trim()) return;
    await addComment(text);
    setText("");
  }

  return (
    <form onSubmit={handleSubmit} className="flex gap-2">
      <input
        type="text"
        placeholder="Write a comment..."
        value={text}
        onChange={(e) => setText(e.target.value)}
        className="flex-1 p-2 border rounded-lg text-sm"
      />
      <button
        type="submit"
        className="bg-blue-500 text-white px-4 py-2 rounded-lg text-sm hover:bg-blue-600"
      >
        Send
      </button>
    </form>
  );
}

export default CommentForm;
