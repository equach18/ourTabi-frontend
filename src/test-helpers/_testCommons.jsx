import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
