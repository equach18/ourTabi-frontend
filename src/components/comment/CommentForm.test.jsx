import { fireEvent, screen, waitFor } from "@testing-library/react";
import CommentForm from "./CommentForm";
import { renderWithAllProviders } from "../../test-helpers/_testCommons";

describe("CommentForm", () => {
  const mockAddComment = vi.fn();

  beforeEach(() => {
    mockAddComment.mockClear();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderWithAllProviders(
      <CommentForm addComment={mockAddComment} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders input and button", () => {
    renderWithAllProviders(<CommentForm addComment={mockAddComment} />);
    expect(screen.getByPlaceholderText(/write a comment/i)).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("calls addComment with text and clears input on submit", async () => {
    renderWithAllProviders(<CommentForm addComment={mockAddComment} />);

    const input = screen.getByPlaceholderText(/write a comment/i);
    const button = screen.getByRole("button");

    fireEvent.change(input, { target: { value: "test comment" } });
    fireEvent.click(button);

    expect(mockAddComment).toHaveBeenCalledWith("test comment");
    await waitFor(() => {
      expect(input.value).toBe("");
    });
  });

  it("does not call addComment if input is empty", () => {
    renderWithAllProviders(<CommentForm addComment={mockAddComment} />);
    fireEvent.click(screen.getByRole("button"));
    expect(mockAddComment).not.toHaveBeenCalled();
  });
});
