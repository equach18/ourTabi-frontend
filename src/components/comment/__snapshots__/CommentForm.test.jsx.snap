// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentForm > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex gap-2"
  >
    <input
      class="flex-1 p-2 border rounded-lg text-sm"
      placeholder="Write a comment..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 text-white px-2 py-2 rounded-lg text-sm hover:bg-blue-600"
      type="submit"
    >
      <svg
        class="size-5"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 12 3.269 3.125A59.769 59.769 0 0 1 21.485 12 59.768 59.768 0 0 1 3.27 20.875L5.999 12Zm0 0h7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;
